floats/floats.go

```diff
@@ -923,3 +923,18 @@ func Within(s []float64, v float64) int {
 	}
 	return -1
 }
+
+// KahanSum returns the sum of the elements of the slice, calculated in a more
+// accurate but also more expensive way. It uses Kahan's compensated summation
+// algorithm: https://en.wikipedia.org/wiki/Kahan_summation_algorithm.
+func KahanSum(s []float64) float64 {
+	sum := 0.
+	c := 0.
+	for _, x := range s {
+		y := x - c
+		t := sum + y
+		c = (t - sum) - y
```

Comment:
@kortschak , is it guaranteed that the compiler will not optimize these calculations away? Does it need to be sparkled with `float64()`?

Reply:
The benchmark timings suggest that it doesn't. If the calculations are optimised away when compiling tests, the test for KahanSum will fail. Do you think that's a sufficient guarantee?

Reply:
@josharian AFAIU `float64` wrapping wouldn't do anything since there's no multiply here. Is that correct?

Reply:
I believe that that is correct, yes. Never hurts to look at the actual generated machine code, though. `GOSSAFUNC=KahanSum go build .` and then look at ssa.html

Reply:
For `NeumaierSum` (a better version of Kahan):

```
genssa
# C:\Users\Roman\go\src\gonum\floats\floats.go
00000 (945) TEXT "".NeumaierSum(SB), ABIInternal
00001 (945) PCDATA $0, $-2
00002 (945) PCDATA $1, $-2
00003 (945) FUNCDATA $0, gclocals·1a65e721a2ccc325b382662e7ffee780(SB)
00004 (945) FUNCDATA $1, gclocals·69c1753bd5f81501d95132d08af04464(SB)
00005 (945) FUNCDATA $2, gclocals·568470801006e5c0dc3947ea998fe279(SB)
v107
00006 (948) PCDATA $0, $0
v107
00007 (948) PCDATA $1, $0
v107
00008 (948) MOVQ "".s+8(SP), AX
v67
00009 (948) PCDATA $0, $1
v67
00010 (948) PCDATA $1, $1
v67
00011 (948) MOVQ "".s(SP), CX
v54
00012 (948) XORL DX, DX
v52
00013 (948) MOVSD $(0.0), X0
v40
00014 (948) MOVSD $(0.0), X1
b1
00015 (948) JMP 18
v98
00016 (+948) INCQ DX
v38
00017 (957) MOVUPS X2, X1
v89
00018 (+948) CMPQ DX, AX
b2
00019 (948) JGE 45
v19
00020 (948) MOVSD (CX)(DX*8), X2
v65
00021 (+949) MOVUPS X2, X3
v21
00022 (949) ADDSD X1, X2
v22
00023 (?) NOP
v51
00024 (?) NOP
v41
00025 (?) NOP
v27
00026 (?) NOP
# $GOROOT\src\math\unsafe.go
v32
00027 (+23) MOVQ X1, BX
# $GOROOT\src\math\abs.go
v36
00028 (+13) BTRQ $63, BX
# $GOROOT\src\math\unsafe.go
v46
00029 (+29) MOVQ BX, X4
v68
00030 (?) NOP
v55
00031 (?) NOP
v60
00032 (+23) MOVQ X3, BX
# $GOROOT\src\math\abs.go
v63
00033 (+13) BTRQ $63, BX
# $GOROOT\src\math\unsafe.go
v73
00034 (+29) MOVQ BX, X5
# C:\Users\Roman\go\src\gonum\floats\floats.go
v45
00035 (+950) UCOMISD X5, X4
b11
00036 (950) JCS 41
v84
00037 (+951) SUBSD X2, X1
v86
00038 (951) ADDSD X1, X3
v87
00039 (951) ADDSD X3, X0
b13
00040 (951) JMP 16
v91
00041 (+953) SUBSD X2, X3
v93
00042 (953) ADDSD X1, X3
v94
00043 (953) ADDSD X3, X0
b14
00044 (953) JMP 16
v101
00045 (+957) PCDATA $0, $0
v101
00046 (+957) ADDSD X0, X1
v104
00047 (957) MOVSD X1, "".~r1+24(SP)
b5
00048 (957) RET
00049 (?) END
```